#!/usr/bin/env python
PACKAGE = "plane_slam"

from dynamic_reconfigure.parameter_generator_catkin import *

gen = ParameterGenerator()

ne_method_enum = gen.enum([gen.const("COVARIANCE_MATRIX", int_t, 0, "COVARIANCE_MATRIX"),
                           gen.const("AVERAGE_3D_GRADIENT", int_t, 1, "AVERAGE_3D_GRADIENT"),
                           gen.const("AVERAGE_DEPTH_CHANGE", int_t, 2, "AVERAGE_DEPTH_CHANGE"),
                           gen.const("SIMPLE_3D_GRADIENT", int_t, 3, "SIMPLE_3D_GRADIENT") ],
                        "An enum to set normal estimation method")

#
gen.add("use_horizontal_line",  bool_t, 0,  "", True)
gen.add("use_verticle_line",  bool_t, 0,  "", True)
gen.add("y_interval",   int_t,  0,  "", 13, 1, 300)
gen.add("x_interval",   int_t,  0,  "", 13, 1, 300)
#
gen.add("line_point_min_distance",   double_t,   0,  "", 0.12, 0.01, 0.3)
gen.add("line_fitting_angular_threshold",    double_t, 0,  "", 3.0, 1.0, 10.0)
gen.add("line_fitting_min_indices",  int_t,   0,  "", 15, 7, 41)
#
gen.add("normals_per_line", int_t,  0,  "", 1,  1,  5)
gen.add("normal_smoothing_size",    int_t,  0,  "", 13, 5, 50)
gen.add("normal_min_inliers_percentage",    double_t,  0,  "", 0.6, 0.2, 0.99)
gen.add("normal_maximum_curvature", double_t,   0,  "", 0.004, 0.0001, 0.008)
#
gen.add("remove_reduplicate_candidate",  bool_t, 0,  "", True)
gen.add("reduplicate_candidate_normal_thresh", double_t,   0,  "", 0.06, 1e-4, 0.5)
gen.add("reduplicate_candidate_distance_thresh", double_t,   0,  "", 0.02, 1e-4, 0.5)
#
gen.add("min_inliers", int_t,  0,  "", 600, 400, 10000)
gen.add("max_curvature", double_t,   0,  "", 0.005, 0.001, 0.5)
gen.add("distance_threshold", double_t,   0,  "", 0.02, 0.001, 0.20)
gen.add("neighbor_threshold",  double_t,   0,  "", 0.2, 0.001, 0.2)
gen.add("angular_threshold",  double_t,   0,  "", 10.0, 1.0, 30.0)
#
gen.add("optimize_coefficients",  bool_t, 0,  "", True)
gen.add("solve_over_segment",  bool_t, 0,  "", True)
gen.add("refine_plane",  bool_t, 0,  "", True)
gen.add("project_points",  bool_t, 0,  "", False)
gen.add("extract_boundary",  bool_t, 0,  "", True)

##
# Normal estimate method:
# - 0: COVARIANCE_MATRIX
# - 1: AVERAGE_3D_GRADIENT
# - 2: AVERAGE_DEPTH_CHANGE
# - 3: SIMPLE_3D_GRADIENT
gen.add("normal_estimate_method", int_t,  3,  "", 0, edit_method=ne_method_enum)
gen.add("normal_estimate_depth_change_factor", double_t,   0,  "", 0.05, 0.001, 0.5)
gen.add("normal_estimate_smoothing_size", int_t,   0,  "", 13, 5, 40)


exit(gen.generate(PACKAGE, "plane_slam", "LineBasedSegment"))
